scripts: sbom: Add GN external project support for Matter

doki-nordic opened the pull request:
The matter project uses GN to build its libraries (not the build system from Zephyr). They are included as CMake ExternalProject. This commit adds detection of such GN ExternalProject and interprets its output ninja file.

This commit also introduces significant source code restructure because ninja build files from GN and Zephyr's CMake files have to be interpreted by one common code.

NordicBuilder commented:
## Test specification

### [CI/Jenkins/NRF](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-9937/4/)
- Integration Platforms

### [CI/Jenkins/integration](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-9937/4/)
| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-9937/4/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-9937
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

peknis commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -548,6 +548,7 @@ This section provides detailed lists of changes by :ref:`script <scripts>`.
 * :ref:`west_sbom`:
 
   * Output now contains source repository and version information for each file.
+  * The script now supports Matter because it can enter GN external projects and get the source files from it.
```
```suggestion
  * The script now supports Matter because it has access to GN external projects and can get the source files from them.
```
them = GN external projects, did I get this right?

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -123,6 +123,14 @@ You can also mix them, for example, to generate a report for the application and
   The :ref:`west_sbom Extracting from build` section describes in detail how to extract a list of files from a build directory.
 
+  You can use ``-d`` option multiple times.
+  For example, to include both the ``mcuboot`` child image and the main application, you can use the following command:
```
```suggestion
  For example, to include both the ``mcuboot`` child image and the main application, use the following command:
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -367,3 +375,10 @@ There are two additional methods for improving the correctness of the above algo
+Integration with the GN meta-build system
+=========================================
+
+The ``ncs-sbom`` reads a list of all commands needed to build provided targets.
```
```suggestion
The ``ncs-sbom`` script reads a list of all commands needed to build provided targets.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -367,3 +375,10 @@ There are two additional methods for improving the correctness of the above algo
+=========================================
+
+The ``ncs-sbom`` reads a list of all commands needed to build provided targets.
+If there is a ``gn gen`` command, the script enters ``gn`` build directory and tries to extract files from it using the same method as described above.
```
```suggestion
If there is a ``gn gen`` command, the script enters the :file:`gn` build directory and tries to extract files from it using the same method as described earlier.
```

Damian-Nordic replied:
I think it's supposed to be "the script enters the build directory used by GN and...". `gn`is not the name of the directory.

doki-nordic replied:
Maybe it can be something like that?
```
If there is a ``gn gen`` command, the script enters the command's build
directory and tries to extract files from it using the same method
as described earlier.
```

doki-nordic replied:
@peknis, what do you think about above sentence?

doki-nordic commented on scripts/west_commands/sbom/ninja_build_extractor.py:
```diff
@@ -0,0 +1,383 @@
+        target_line_re = re.compile(r'([^\s]+)\s*:\s*(#.*)?')
+        dep_line_re = re.compile(r'\s+(.*?)\s*(#.*)?')
+        empty_line_re = re.compile(r'\s*(#.*)?')
+        with open(deps_file_name, 'r') as fd:
```
For `detect_file_type` function, encoding is required, because the file that we are opening there may be binary, so we have to force encoding that will not raise any exceptions for binary files. This case is different, we are opening text only file using platform's default encoding and it is ok to raise an exception if there is something wrong with the file.

doki-nordic commented on scripts/west_commands/sbom/ninja_build_extractor.py:
```diff
@@ -0,0 +1,383 @@
+        if header.startswith('!<arch>\n'):
+            self.file_type_cache[str_file_name] = FileType.ARCHIVE
+            return FileType.ARCHIVE
+        elif header.startswith('\x7FELF'):
```
It is sufficient, but using chain of `if-elif` is much more readable. You don't need to analyze body of previous blocks to see general flow. That is why I prefer to keep it this way.

doki-nordic commented on scripts/west_commands/sbom/ninja_build_extractor.py:
```diff
@@ -0,0 +1,383 @@
+        the specified targets. The returned value is in a BuildArchive object.
+        '''
+        def process_target_execute(exec_args):
+            self.lock.acquire()
+            self.process_target(*exec_args)
+            self.lock.release()
```
Good catch. The `acquire/release` implementation cause dead-lock that I hunted to some time.

doki-nordic commented:
I rebased and updated the PR with some significant improvements:
* I am querying ninja with multiple targets at once. Manual does not clearly states that ninja allows `-t query` command for multiple targets, but looks like it allows it and it is implemented in ninja from the beginning, so I think it is pretty safe to do that. It causes huge performance boost.
* I fixed mutex locking code, so it is immune to deadlocks now.
* I fixed unnecessary files in the report, e.g. `libCHIP.a` should not appear in the report when all its sources are already there.
* Small code and docs fixes based on rewiev.

You can see updates here:
https://github.com/nrfconnect/sdk-nrf/compare/96f6985343001e55cb055ddcc988f6327cbe681a..a18a64466e5ecd931944df0876c1e295439a6b9a

maje-emb reviewed (approved):
Some minor comments.
